libs/ui/src/hooks/smartcard_auth/auth_helpers.test.ts

```diff
@@ -0,0 +1,226 @@
+
+    jest.spyOn(cardApi, 'writeLongUint8Array').mockRejectedValue(error);
+    expect(
+      (await result.current.card.writeStoredData(Uint8Array.of(1, 2, 3))).err()
+    ).toEqual(error);
+
+    expect((await result.current.card.clearStoredData()).err()).toEqual(error);
+  });
+
+  it('raises an error on concurrent writes', async () => {
+    const cardApi = new MemoryCard();
+    cardApi.insertCard(makePollWorkerCard(electionHash));
+    const { result, waitForNextUpdate } = renderHook(() =>
+      useInsertedSmartcardAuth({
+        cardApi,
+        allowedUserRoles,
+        scope: { electionDefinition },
+      })
+    );
+    jest.advanceTimersByTime(CARD_POLLING_INTERVAL);
+    await waitForNextUpdate();
+    assert(isPollworkerAuth(result.current));
+
+    const [write1, write2] = await Promise.all([
+      result.current.card.writeStoredData(electionDefinition),
+      result.current.card.writeStoredData(electionDefinition),
+    ]);
+
+    expect(write1.isErr()).toBe(false);
+    expect(write2.isErr()).toBe(true);
+    expect(write2.err()?.message).toEqual('Card write in progress');
+
+    const [write3, clear4] = await Promise.all([
+      result.current.card.writeStoredData(electionDefinition),
+      result.current.card.clearStoredData(),
+    ]);
+    expect(write3.isErr()).toBe(false);
+    expect(clear4.isErr()).toBe(true);
+    expect(clear4.err()?.message).toEqual('Card write in progress');
+
+    // Make sure failed writes still release the lock
+    jest
+      .spyOn(cardApi, 'writeLongObject')
+      .mockRejectedValue(new Error('error'));
+    expect(
+      (await result.current.card.writeStoredData(electionDefinition)).isErr()
+    ).toBe(true);
+    expect((await result.current.card.clearStoredData()).isOk()).toBe(true);
+  });
+});
+
+describe('Type guards', () => {
+  test('isSuperadminAuth', () => {
+    expect(isSuperadminAuth(Inserted.fakeSuperadminAuth())).toBe(true);
+    expect(isSuperadminAuth(Inserted.fakeVoterAuth())).toBe(false);
+    expect(isSuperadminAuth(Inserted.fakeLoggedOutAuth())).toBe(false);
+    expect(isSuperadminAuth(Dipped.fakeSuperadminAuth())).toBe(true);
+    expect(isSuperadminAuth(Dipped.fakeAdminAuth())).toBe(false);
+    expect(isSuperadminAuth(Dipped.fakeLoggedOutAuth())).toBe(false);
+  });
+
+  test('isAdminAuth', () => {
+    expect(isAdminAuth(Inserted.fakeAdminAuth())).toBe(true);
+    expect(isAdminAuth(Inserted.fakeVoterAuth())).toBe(false);
+    expect(isAdminAuth(Inserted.fakeLoggedOutAuth())).toBe(false);
+    expect(isAdminAuth(Dipped.fakeAdminAuth())).toBe(true);
+    expect(isAdminAuth(Dipped.fakeSuperadminAuth())).toBe(false);
+    expect(isAdminAuth(Dipped.fakeLoggedOutAuth())).toBe(false);
+  });
+
+  test('isPollworkerAuth', () => {
+    expect(isPollworkerAuth(Inserted.fakePollworkerAuth())).toBe(true);
+    expect(isPollworkerAuth(Inserted.fakeVoterAuth())).toBe(false);
+    expect(isPollworkerAuth(Inserted.fakeLoggedOutAuth())).toBe(false);
+  });
```

Comment:
Dipped auth only supports admins and superadmins, and I believe I set the types up so that it's a type error to even call `isPollworkerAuth` on dipped auth.